tx-parser/src/lib.rs

```diff
@@ -39,165 +42,114 @@ use mentat_tx::entities::{
-        .into()
+def_parser!(Tx, atom, edn::ValueAndSpan, {
+    satisfy_map(|x: edn::ValueAndSpan| x.into_atom().map(|v| v))
 });
 
-fn value_to_add_or_retract(val: &Value) -> Option<Entity> {
-    val.as_vector().and_then(|vs| {
-        let add = token(Value::NamespacedKeyword(NamespacedKeyword::new("db", "add")))
-            .map(|_| OpType::Add);
-        let retract = token(Value::NamespacedKeyword(NamespacedKeyword::new("db", "retract")))
-            .map(|_| OpType::Retract);
-        let mut p = (add.or(retract),
-                     Tx::<&[Value]>::entid_or_lookup_ref_or_temp_id(),
-                     Tx::<&[Value]>::entid(),
-                     Tx::<&[Value]>::atom_or_lookup_ref_or_vector())
-            .skip(eof())
-            .map(|(op, e, a, v)| {
-                Entity::AddOrRetract {
-                    op: op,
-                    e: e,
-                    a: a,
-                    v: v,
-                }
-            });
-        let r: ParseResult<Entity, _> =  p.parse_lazy(&vs[..]).into();
-        r.map(|x| x.0).ok()
-    })
-}
-def_value_satisfy_parser_fn!(Tx, add_or_retract, Entity, value_to_add_or_retract);
+def_parser!(Tx, nested_vector, Vec<AtomOrLookupRefOrVectorOrMapNotation>, {
+    vector().of(many(Tx::atom_or_lookup_ref_or_vector()))
+});
 
-fn value_to_map_notation(val: &Value) -> Option<MapNotation> {
-    val.as_map().cloned().and_then(|map| {
-        // Parsing pairs is tricky; parsing sequences is easy.
-        let avseq: Vec<Value> = map.into_iter().flat_map(|(a, v)| once(a).chain(once(v))).collect();
+def_parser!(Tx, atom_or_lookup_ref_or_vector, AtomOrLookupRefOrVectorOrMapNotation, {
+    Tx::lookup_ref().map(AtomOrLookupRefOrVectorOrMapNotation::LookupRef)
+        .or(Tx::nested_vector().map(AtomOrLookupRefOrVectorOrMapNotation::Vector))
+        .or(Tx::map_notation().map(AtomOrLookupRefOrVectorOrMapNotation::MapNotation))
+        .or(Tx::atom().map(AtomOrLookupRefOrVectorOrMapNotation::Atom))
+});
 
-        let av = (Tx::<&[Value]>::entid(),
-                  Tx::<&[Value]>::atom_or_lookup_ref_or_vector())
-            .map(|(a, v)| -> (Entid, AtomOrLookupRefOrVectorOrMapNotation) { (a, v) });
-        let mut p = many(av)
-            .skip(eof())
-            .map(|avs: Vec<(Entid, AtomOrLookupRefOrVectorOrMapNotation)>| -> MapNotation {
-                avs.into_iter().collect()
-            });
-        let r: ParseResult<MapNotation, _> =  p.parse_lazy(&avseq[..]).into();
-        r.map(|x| x.0).ok()
-    })
-}
-def_value_satisfy_parser_fn!(Tx, map_notation, MapNotation, value_to_map_notation);
+def_parser!(Tx, add_or_retract, Entity, {
+    let add = value(edn::Value::NamespacedKeyword(NamespacedKeyword::new("db", "add")))
+        .map(|_| OpType::Add);
+    let retract = value(edn::Value::NamespacedKeyword(NamespacedKeyword::new("db", "retract")))
+        .map(|_| OpType::Retract);
+    let p = (add.or(retract),
+             Tx::entid_or_lookup_ref_or_temp_id(),
```

Comment:
![it s beautiful](https://cloud.githubusercontent.com/assets/91722/24530910/f79abafa-1569-11e7-9179-837a9ab7aec0.jpg)